dali/pipeline/executor/executor.cc

```diff
@@ -398,6 +399,38 @@ void Executor<WorkspacePolicy, QueuePolicy>::RunHelper(OpNode &op_node, Workspac
                              ws.GetRequestedBatchSize(i), " <= ", max_batch_size_));
   }
 
+  // If we are using conditional statements try to keep smaller memory footprint by not keeping
+  // the largest batch produced in conditional block as reserved memory.
+  // If the conditionals were detected in this graph (split operator with `_if_stmt` argument)
+  // see if the operator is executed in conditional scope - seeing a partial batch processed
+  // would indicate it in most cases (unless someone mixes partial batches and conditionals).
+  // In such case reset the outputs before adjusting their size so we free the reserved memory
+  // and allocated only batch of currently needed size.
+  if (has_conditionals_) {
```

Comment:
So if the condition is applied to any part of the graph we reset buffer for all workspaces, right?

Reply:
`has_conditionals_` is global for the whole graph.
Here we check `ws.GetInputBatchSize(i) < max_batch_size_` for all ops in the `has_conditionals_` is set, not only for the ops inside the scope.
Am I right?

Reply:
>   // If the conditionals were detected in this graph (split operator with `_if_stmt` argument)
  // see if the operator is executed in conditional scope - seeing a partial batch processed

If there are any conditionals in the graph I treat an op processing a batch smaller than max_batch_size as running in a conditional scope - as this is typically a good indication, apart for some edge cases.

We can extend it by adding proper scoping information to every operator, but that is more involved change and I am not sure how to make it in a good, forward compatible way.
My current idea is to add a hidden parameter with the predicate to every operator in conditional scope, but that just adds a lot of edges in graph. Another idea is to number every scope and pass those ids around so we can look it up. This seems a bit better, but would be a bigger change.

Reply:
I think that long term it would be to move to a fully dynamic allocation of the memory for each workspace, each iteration, and limit the liveness of the buffers. For now, I would stay with the proposed approach and don't try harder with it.

Reply:
I am not sure what we are discussing here any more. This basically does dynamic allocation when there are conditionals in the graph for anything that looks like running in a conditional scope. If there are no conditionals or something is processing a max_batch_size, we keep the current behaviour.

Reply:
I'm just trying to understand how it works and if it is aligned with the comment.